### PR TITLE
Create projects from manifests

### LINKED ISSUE #1852: Ability to create & upsert projects from spec
Similar to `argocd app create -f /app-spec.yaml`, we need equivalent capability to create a project from spec.

e.g. 
```
argocd proj create -f /proj-spec.yaml
```

NOTE: this API should support --upsert mode.

### PR DESCRIPTION
**Pull request by Rayyis**
This commit allows users to create projects from manifests & upsert them through the CLI or rest API.
It closes #1852 

---
Changes roundup:
- Added an option to upsert projects
- Added an option to create projects from manifests
- Both apps and projects can read manifests from the stdin using `-f -` similar to kubectl
- Modified the e2e project creation test to cover the changes above

### CODE REVIEW

File: cmd/argocd/commands/project.go [hunk 1]
```diff
@@ -120,33 +122,53 @@ func humanizeTimestamp(epoch int64) string {
...
-				ObjectMeta: v1.ObjectMeta{Name: projName},
-				Spec: v1alpha1.AppProjectSpec{
-					Description:       opts.description,
-					Destinations:      opts.GetDestinations(),
-					SourceRepos:       opts.sources,
-					OrphanedResources: getOrphanedResourcesSettings(c, opts),
-				},
-			}
+			var proj v1alpha1.AppProject
 			conn, projIf := argocdclient.NewClientOrDie(clientOpts).NewProjectClientOrDie()
```
Comment (alexec) [inline cmd/argocd/commands/project.go]: do you need to get this on this line, can you wait until line 153?

File: cmd/argocd/commands/project.go [hunk 2]
```diff
@@ -120,33 +122,53 @@ func humanizeTimestamp(epoch int64) string {
...
+			if fileURL != "" {
+				parsedURL, err := url.ParseRequestURI(fileURL)
+				if err != nil || !(parsedURL.Scheme == "http" || parsedURL.Scheme == "https") {
+					err = config.UnmarshalLocalFile(fileURL, &proj)
+				} else {
+					err = config.UnmarshalRemoteFile(fileURL, &proj)
+				}
+				errors.CheckError(err)
+				if len(args) == 1 && args[0] != proj.Name {
+					log.Fatalf("project name '%s' does not match project spec metadata.name '%s'", args[0], proj.Name)
```
Comment (alexec) [inline cmd/argocd/commands/project.go]: good check

**Review by alexec — COMMENTED** [review]
This looks good. I think it could probably do with an e2e test. what do you think?

File: cmd/argocd/commands/project.go [hunk 3]
```diff
@@ -120,33 +122,53 @@ func humanizeTimestamp(epoch int64) string {
...
+						SourceRepos:  opts.sources,
+					},
+				}
+			}
+			_, err := projIf.Create(context.Background(), &projectpkg.ProjectCreateRequest{Project: &proj, Upsert: upsert})
 			errors.CheckError(err)
 		},
 	}
+	command.Flags().BoolVar(&upsert, "upsert", false, "Allows to override a project with the same name even if supplied project spec is different from existing spec")
+	command.Flags().StringVarP(&fileURL, "file", "f", "", "Filename or URL to Kubernetes manifests for the project")
```
[the remark this answers is not in the document]
Comment (Rayyis) [reply cmd/argocd/commands/project.go]: Good point, I'll modify this and the app create command to read from stdin

File: cmd/argocd/commands/project.go (hunk 1 above)
Comment (Rayyis) [reply cmd/argocd/commands/project.go]: It is needed for unmarshaling in 139 and 141, though I noticed the connection below it could wait a bit.

**Comment by Rayyis** [comment]
Thanks for the feedback @alexec, Could you take another look?

**Review by alexec — COMMENTED** [review]
Only one suggestion, that `RunWithStdin` take a string for the stdin. Thoughts?

**Review by alexec — COMMENTED** [review]
This looks good to me. @alexmt or @jessesuen fancy a look as this is about projects.

File: test/e2e/project_management_test.go [hunk 4]
```diff
@@ -66,6 +69,28 @@ func TestProjectCreation(t *testing.T) {
...
+
+	// create a manifest with the same name to upsert
+	newDescription := "Upserted description"
+	proj.Spec.Description = newDescription
+	proj.ResourceVersion = ""
+	data, err := json.Marshal(proj)
+	assert.NoError(t, err)
+
+	// fail without upsert flag
+	var reader io.Reader = bytes.NewReader(data)
```
[the remark this answers is not in the document]
Comment (Rayyis) [reply test/e2e/project_management_test.go]: Yeah it'd make more sense to create readers later from the string

**Review by alexmt — APPROVED** [review]
LGTM!

**Review by alexec — APPROVED** [review]

**Pull request merged**